util/src/main/java/com/google/cloud/hadoop/util/ThreadTrace.java

```diff
@@ -0,0 +1,186 @@
+
+  public ThreadTrace(String trackingId) {
+    this(trackingId, new ArrayList<>());
+  }
+
+  private ThreadTrace(String trackingId, List<ThreadTraceEvent> events) {
+    checkNotNull(trackingId, "trackingId cannot be null");
+    checkNotNull(events, "events cannot be null. trackingId=%s", trackingId);
+    this.trackingId = trackingId;
+    this.events = events;
+  }
+
+  @Override
+  public String toString() {
+    return gson.toJson(this);
+  }
+
+  ThreadTraceEvent starOperationEvent(String name) {
+    if (isFull()) {
+      return null;
+    }
+
+    ThreadTraceEvent theEvent = ThreadTraceEvent.startEvent(name);
+    this.events.add(theEvent);
+
+    return theEvent;
+  }
+
+  void endOperationEvent(ThreadTraceEvent startEvent) {
+    if (isFull() || startEvent == null) {
+      return;
+    }
+
+    int eventsSize = events.size();
+    if (eventsSize != 0) {
+      ThreadTraceEvent lastEvent = events.get(eventsSize - 1);
+      if (lastEvent == startEvent) {
+        startEvent.markEnd();
+        return;
+      }
+    }
+
+    this.events.add(startEvent.endEvent());
+  }
+
+  public String getTrackingId() {
+    return this.trackingId;
+  }
+
+  ThreadTrace getSubTrace() {
+    long threadId = Thread.currentThread().getId();
+
+    subEvents.putIfAbsent(threadId, new ArrayList<>());
+
+    int size = subEvents.get(threadId).size();
+    if (size >= MAX_EVENTS_SIZE) {
+      return null;
+    }
+
+    return new ThreadTrace(this.trackingId, subEvents.get(threadId).subList(size, size));
+  }
+
+  public void annotate(String key, Object value) {
+    if (this.events.size() == 0) {
+      return;
+    }
+
+    this.events.get(this.events.size() - 1).annotate(key, value);
+  }
+
+  private boolean isFull() {
+    return events.size() >= MAX_EVENTS_SIZE;
+  }
+
+  static class ThreadTraceEvent extends HashMap<String, Object> {
+
+    static final int MAX_PROPERTY_SIZE = 10;
+
+    private enum EventType {
+      START,
+      END,
+      // This indicates that two consecutive start-end event pairs are "merged" to one event. This
+      // is done to reduce the number of events in the trace.
+      MERGED,
+    }
+
+    @Expose(serialize = false)
+    private final long timeStamp;
+
+    @Expose(serialize = false)
+    private final String name;
+
+    static ThreadTraceEvent startEvent(String name) {
+      return new ThreadTraceEvent(EventType.START, name);
+    }
+
+    private ThreadTraceEvent(EventType type, String name) {
+      super(5);
+      this.setEventType(type);
+      this.put("name", name);
+      this.timeStamp = System.currentTimeMillis();
+      this.name = name;
+      if (type == EventType.START) {
+        this.put("starttime", this.timeStamp);
+      } else if (type == EventType.END) {
+        this.put("endtime", this.timeStamp);
+      } else {
+        throw new IllegalStateException("Not expected to create event of type: " + type);
+      }
+    }
+
+    public void markEnd() {
```

Comment:
1. shouldn't we verify that `this` is of type `START` before updating the type to `MERGED`?
2. Shouldn't we capture the `endtime` as well just for completeness purposes and maybe pass the `endTime` as an argument to this function.

Reply:
1. This will always be start since it is created in the constructor.
2. If we use endtime, we still need to subtract to find the time taken. We are more interested in the time taken by the operation. endTime can be derived from startTime and timeTaken. There is no value in duplicating the same information.